test/test_while_loop.py

```diff
@@ -0,0 +1,182 @@
+    _, expected = _fake_while_loop(cond_fn, body_fn, (iteri, init_val))
+    print("res: ", res)
+    print("expected: ", expected)
+    self.assertTrue(torch.all(torch.eq(res, expected)))
+
+  def test_while_loop_tpu_addition_nested(self):
+    device = xm.xla_device()
+
+    def cond_fn(iteri, x):
+      return iteri > 0
+
+    def body_fn(iteri, x):
+      return iteri - 1, torch.add(torch.add(x, 1), 1)
+
+    init_val = torch.tensor(2, dtype=torch.int32, device=device)
+    iteri = torch.tensor(10, device=device)
+    _, res = while_loop(cond_fn, body_fn, (iteri, init_val))
+    _, expected = _fake_while_loop(cond_fn, body_fn, (iteri, init_val))
+    print("res: ", res)
+    print("expected: ", expected)
+    self.assertTrue(torch.all(torch.eq(res, expected)))
+
+  def test_while_loop_tpu_simple_linear_inside_loop(self):
+    device = xm.xla_device()
+    torch.set_grad_enabled(False)
+
+    class SimpleLinear(torch.nn.Module):
+
+      def __init__(self):
+        super().__init__()
+        self.linear = torch.nn.Linear(2, 2)
+
+      def forward(self, iteri, x):
+
+        def cond_fn(iteri, x):
+          return iteri > 0
+
+        def body_fn(iteri, x):
+          return iteri - 1, self.linear(x)
+
+        return while_loop(cond_fn, body_fn, (iteri, x))
+
+      def forward_compare(self, iteri, x):
+        y = self.linear(x)
+        return iteri - 1, y
+
+    linear_model = SimpleLinear()
+    linear_model.to(device)
+    l_in_0 = torch.randn(2, 2, dtype=torch.float32, device=device)
+    iteri = torch.tensor(2, dtype=torch.int32, device=device)
+    _, res = linear_model(iteri, l_in_0)
+
+    # === expected result after 2 iteration to be compared ===
+    _, expected = linear_model.forward_compare(iteri, l_in_0)
+    _, expected = linear_model.forward_compare(iteri, expected)
+
+    print("res: ", res)
+    print("expected: ", expected)
+    self.assertTrue(torch.all(torch.eq(res, expected)))
+
+  def test_while_loop_tpu_MNIST_inside_loop(self):
+    device = xm.xla_device()
+    torch.set_grad_enabled(False)
+
+    class MNIST(torch.nn.Module):
+
+      def __init__(self):
+        super().__init__()
+        self.conv1 = torch.nn.Conv2d(1, 10, kernel_size=5, stride=1, padding=2)
+        self.bn1 = torch.nn.BatchNorm2d(10).eval()
+        self.conv2 = torch.nn.Conv2d(10, 20, kernel_size=5)
+        self.bn2 = torch.nn.BatchNorm2d(20).eval()
+        self.fc1 = torch.nn.Linear(500, 50)
+        self.fc2 = torch.nn.Linear(50, 10)
+        self.bnLayersWeights = []
+
+      def forward(self, iteri, x, y):
+
+        def cond_fn(iteri, x, y):
+          return iteri > 0
+
+        def body_fn(iteri, x, y):
+          y = F.relu(F.max_pool2d(self.conv1(x), 2))
+          y = self.bn1(y)
+          y = F.relu(F.max_pool2d(self.conv2(y), 2))
+          y = self.bn2(y)
+          y = torch.flatten(y, 1)
+          y = F.relu(self.fc1(y))
+          y = self.fc2(y)
+
+          return iteri - 1, x.clone(), F.log_softmax(y, dim=1)
```

Comment:
I am confuse, wouldn't `x` always be the same in this case? Shouldn't you return `F.log_softmax(y, dim=1)` as x?

Reply:
thanks, we add two test case for two situation:
`test_while_loop_tpu_simple_linear_inside_loop` is the test case that we return y as next x, due to y and x has the same shape and type;
`test_while_loop_tpu_MNIST_inside_loop` is the test case that we return x and y both, due to y and x has differeny shape and type;

Reply:
Thanks, updated test without MNIST in this PR, would add MNIST test case in the next PR